Python template v3

albireox opened the pull request:
I have done some updating of both the template and the style guide, which were over three years old.

This new template (v3) uses [copier](https://copier.readthedocs.io/en/stable/) because `cookiecutter` seems to have become abandonware.  I have remove most of the configuration options and settled for using uv, Ruff, and furo for Sphinx theming since I think that's what we all are using these days and it simplifies the template and the documentation.

The new version of the documentation can be previewed [here](https://sdss-python-template.readthedocs.io/en/python-template-v3/index.html) and the style guide [here](https://sdss-python-template.readthedocs.io/en/python-template-v3/style/STYLE_v2.html).

This probably deserves a synchronous code-review.

Closes #25

albireox committed:
Mostly blank slate for v3 with stub documentation

albireox committed:
Initial template

albireox committed:
Add licenses

albireox committed:
Support downloading licenses and other improvements

albireox committed:
Add src/python_project to allow Sphinx to find the package

albireox committed:
Do not ask to keep Copier answers

albireox committed:
Add .vscode to .gitignore

albireox committed:
Use uv sync --frozen --group test

albireox committed:
60% of documentation

albireox committed:
Add pytest and coverage configuration to pyproject.toml

albireox committed:
Add section on scripts to docs

albireox committed:
More or less complete documentation

albireox committed:
Update style guide to v2

albireox committed:
Update documentation and add badges to README

albireox committed:
Update badges

albireox committed:
Import package earlier in test

albireox committed:
Do not omit files for coverage

albireox committed:
Update README

albireox committed:
Update changelog

johndonor3 reviewed (approved):
I don't have any strong opinions here. It seems simpler than the previous template though so that's good!

havok2063 commented:
I'll take a look at this soon.  Maybe by next week's coding time.  Regarding `copier`, is `cookiecutter` deprecated?  it's last commit was 5 days ago.  And I know folks are using [cruft](https://cruft.github.io/cruft/) now to manage project templates.  I'm not familiar with `copier` but I'll take a look.

albireox commented:
Yes, it seems people still commit to `coockiecutter` but there hasn't been a release in over a year. I checked the issues at the time and it seemed the main maintainers where MIA, but maybe not. I didn't know cruft but I'm not super attached to copier either. I think any of them are probably good enough for our needs ... including an old version of cookiecutter.

havok2063 commented:
I'm testing this out.  When I run `uvx copier copy --trust gh:sdss/python_template /Users/brian/Work/git/sdss/test_project` I get an error.  I get the same thing pointing to my local checkout of the branch.  This shouldn't depend on `cookiecutter`, right?
```
  File "<template>", line 1, in top-level template code
  File "/Users/brian/.local/share/uv/tools/copier/lib/python3.10/site-packages/jinja2/sandbox.py", line 319, in getattr
    value = getattr(obj, attribute)
jinja2.exceptions.UndefinedError: 'cookiecutter' is undefined
```

albireox commented:
I think that's because it's defaulting to the `main` branch which still has references to `cookiecutter` since this is not yet merged. Can you try `uvx copier copy --trust --vcs-ref python-template-v3 gh:sdss/python_template /Users/brian/Work/git/sdss/test_project`?

havok2063 commented:
That worked.  I'm surprised it didn't work when it was pointed to the local branch checkout.  

One quick question on the uv venv setup:  What determines which python version it creates the venv for?  It defaults to 12, whether I created it from a different env or not. Are we able to have a question asking which python version to use?  

I'll keep playing around with this.

albireox commented:
I think to run it from a local checkout you need to do something like `copier copy --trust -r HEAD . ~/Downloads/test_project` The `-r HEAD` was necessary if you have uncommitted changes.

Yes, we could ask a question about the `uv` Python version to use. Right now it defaults to whatever `uv` would use if you do `uv venv`, which I think is the highest version it has access to, but there are environment variables to change that.

albireox commented:
Yesterday you mentioned being able to update the template for already generated project. I think that's a good idea. `copier` has [some features](https://copier.readthedocs.io/en/stable/updating/) for that that I haven't explored. I think they would require keeping the answers file, which right now I'm removing in post-processing.

havok2063 commented on template/README.md.jinja:
```diff
@@ -0,0 +1,17 @@
+[![License](https://img.shields.io/badge/License-GPLv3-blue.svg)](https://www.gnu.org/licenses/gpl-3.0)
+{% endif %}
+<!-- [![Documentation Status](https://readthedocs.org/projects/{{ project_name }}/badge/?version=latest)](https://{{ project_name }}.readthedocs.io/en/latest/) -->
+<!-- [![codecov](https://codecov.io/gh/{{github_organization}}/{{ project_name }}/branch/main/graph/badge.svg)](https://codecov.io/gh/{{github_organization}}/{{ project_name }}) -->
```
Can we add some boilerplate instructions to this readme for installation, dev installation, and how to serve the docs locally?

albireox replied:
Done. Can you check the text? I mostly copied what you added for `sdssdb`.

havok2063 replied:
It looks like it was added to the main python_template readme.  I was actually thinking about adding stub text to the template here so built projects already had that info in their readme files.

havok2063 commented on template/docs/sphinx/conf.py.jinja:
```diff
@@ -0,0 +1,143 @@
+    "sphinx.ext.mathjax",
+    "sphinx.ext.intersphinx",
+    "myst_parser",
+    "sphinx_copybutton",
```
Can we add `sphinx_inline_tabs` as an added extension?  I find this a useful extension that might be nice to include.

albireox replied:
Added,

havok2063 reviewed (commented):
This is looking good to me.  The output project is nice and clean.  One thought I had was adding a question to optionally create a minimal Dockerfile for the package.  

Also, I wonder if it's worth thinking about a migration script to help packages move from v1->v2->v3, etc, or future versions.  I'm not sure how feasible that is.  Perhaps it's easier moving forward, after projects are made with `copier`.

havok2063 commented:
Yeah I think "project updates" would be good to include.  Besides supporting package/project updates, it would allow people to re-answer optional questions for adding features.  

We might also want a separate script/repo for pushing mass updates to multiple packages.

albireox committed:
Add sphinx_inline_tabs

albireox committed:
Add Makefile for Sphinx documentation

albireox committed:
Add instructions for local development and building the docs to README.md

albireox committed:
Add link to documentation in README

albireox committed:
Add an option to include a Dockerfile and GitHub Action workflow

albireox commented:
I included a question to add a `Dockerfile` and GitHub action. I tested it and I think it works. Can you have a look at the code and the documentation?

I like the idea of allowing to update projects generated with the template. I will probably require a bit of testing and playing around with the options.

I think an automated conversion tool from v1 and v2 may be difficult to write since those versions had some many forking paths, but maybe we can include some documentation on how to do it manually.

albireox committed:
Allow to not delete the .copier-answers.yaml file

albireox committed:
Add some files to path-ignore

albireox committed:
Always keep answers

albireox committed:
Add docs on how to update the project

havok2063 commented:
This is looking good.  The Docker addition seems reasonable to me, and the docs look fine.   If we're coding today, we can touch base on this.  I'd also like to trial and test the project update, but that can also be done outside this PR, if we want.

albireox commented:
I was travelling this week and wasn't able to call to the coding meeting. Let's discuss it next Thursdays and maybe merge. We can test the update system and confirm it works.

albireox committed:
Update Dockerfile to use Python 3.14

albireox committed:
Reverting back to Python 3.13 in the Dockerfile

albireox committed:
Add question about minimum Python version

albireox committed:
Optionally add framework dependencies

albireox committed:
Add project description to GitHub repo

albireox committed:
Add instructions for managing the project, coding standards, and building the documentation to the README template

albireox commented:
I think I addressed the things we discussed today:

- The questionnaire ask for a minimum Python version which is used in `pyproject.toml`, the `Dockerfile`, and to determine the test matrix.
- The questionnaire asks whether to include framework packages. `sdss-access`, `sdss-tree`, and `sdssdb` are added if so.
- Added some instructions on how to start using the rendered template to the `README.md`.

Please have a look and let me know how those look.

albireox committed:
Update questionnaire in docs

albireox committed:
Make package_name default to project_name with replacements

albireox committed:
Tweak question text

albireox committed:
A few other question tweaks

albireox committed:
Update docs with new questions in copier.yaml

albireox committed:
Do not try to push repo if GitHub username is not provided

havok2063 commented:
I tested this.  I found one issue with this question "🐍 The minimum Python version required to use the project (>= 3.8).
   3.13".  I specified to use 3.13 but the version of Python that got installed in the venv was 3.14.

havok2063 commented on template/README.md.jinja:
````diff
@@ -0,0 +1,69 @@
+If you want tos specify a different Python version for the virtual environment, you can use the `--python` flag with `uv venv`, e.g.
+
+```console
+uv venv --python 3.11
````
```suggestion
uv venv --python {{ minimum_python_version }}
```

albireox replied:
Can you try now? I would leave the documentation example as it is, but I modified the post_copy script to force uv to use the minimum python version from the answers.

albireox committed:
Create the virtual environment using the questionnaire Python version

havok2063 commented on template/post_copy.py:
```diff
@@ -0,0 +1,286 @@
+                    "UV_ACTIVE=0",
+                    f"UV_PROJECT_ENVIRONMENT={cwd!s}/.venv",
+                    f"VIRTUAL_ENV={cwd!s}/.venv",
+                    "UV_PYTHON=''",
```
Should this be changed to `UV_PYTHON={{ minimum_python_version }}"`?

albireox replied:
Yep, that's the change I made.

havok2063 reviewed (approved):
Looks good.  The min python version now works.
